libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp

```diff
@@ -155,56 +107,120 @@ void SortitionParamsManager::pbftBlockPushed(const SyncBlock& block, DbStorage::
+      const auto params_change = calculateChange(period);
+      params_changes_.push_back(params_change);
+      db_->saveSortitionParamsChange(period, std::move(params_change), batch);
+      cleanup();
     }
+  } else {
+    ignored_efficiency_counter_++;
+  }
+}
 
-    cleanup(period);
+int32_t efficiencyToChange(uint16_t efficiency, uint16_t goal_efficiency) {
+  uint16_t deviation = std::abs(efficiency - goal_efficiency) * 100 / goal_efficiency;
+  // If goal is 50 % return 1% change for 40%-60%, 2% change for 30%-70% and 5% change for over that
+  if (deviation < 20) {
+    return UINT16_MAX / 100;
   }
+  if (deviation < 40) {
+    return UINT16_MAX / 50;
+  }
+  return UINT16_MAX / 20;
 }
 
-int32_t SortitionParamsManager::getChange(uint64_t period, uint16_t efficiency) const {
-  const uint16_t per_percent = averageCorrectionPerPercent();
-  int32_t correction = -(config_.targetEfficiency() - efficiency);
-  if (std::abs(correction) > config_.max_interval_correction) {
-    correction = (correction < 0 ? -1 : 1) * config_.max_interval_correction;
+int32_t SortitionParamsManager::getNewUpperRange(uint16_t efficiency) const {
+  assert(params_changes_.size() > 0);
+  if (efficiency >= config_.dag_efficiency_targets.first && efficiency <= config_.dag_efficiency_targets.second) {
+    return params_changes_[params_changes_.size() - 1].vrf_params.threshold_upper;
   }
-  const int32_t change = correction * per_percent / kOnePercent;
 
-  LOG(log_dg_) << "Average interval efficiency: " << efficiency / 100. << "%, correction per percent: " << per_percent;
-  LOG(log_nf_) << "Changing VRF params on " << period << " period from (" << config_.vrf.threshold_range << ", "
-               << config_.vrf.threshold_upper << ") by " << change;
+  // efficiencies_to_uppper_range provide mapping from efficiency to VRF upper threshold, params_changes contain
+  // efficiency for previous setting so mapping is done efficiency of i relates to VRF upper threshold of (i + 1)
+  std::map<uint16_t, uint32_t> efficiencies_to_uppper_range;
+  const uint16_t goal_efficiency = (config_.dag_efficiency_targets.first + config_.dag_efficiency_targets.second) / 2;
+  for (uint32_t i = 1; i < params_changes_.size(); i++) {
+    efficiencies_to_uppper_range[params_changes_[i].interval_efficiency] =
+        params_changes_[i - 1].vrf_params.threshold_upper;
+  }
+  if (params_changes_.size() > 1) {
+    efficiencies_to_uppper_range[efficiency] = params_changes_[params_changes_.size() - 1].vrf_params.threshold_upper;
+  }
 
-  return change;
-}
+  // Check if all last params are below goal_efficiency
+  if ((efficiencies_to_uppper_range.empty() || efficiencies_to_uppper_range.rbegin()->first < goal_efficiency) &&
+      efficiency < goal_efficiency) {
+    // If last params are under goal_efficiency and we are still under goal_efficiency, decrease upper
+    // limit
+    return ((int32_t)params_changes_[params_changes_.size() - 1].vrf_params.threshold_upper) -
```

Comment:
static_cast 
Also I think we need to avoid this much operations in one line. It takes too much time to get what is going on here. Comment clears it much, but we can split it and make readable without it

Reply:
I have now reduced these long lines.